### ISSUE TITLE
Error in Windows 10 LTSC  x64

### ISSUE THREAD

**Issue by gocict** [basil00/Divert#196]
I run Windivert on Windows 10 x64 and it has this error

![Untitled](https://user-images.githubusercontent.com/46121759/59145339-80ffd000-8a0c-11e9-8f65-3db22befbfd9.png)

Is this a bug?

**Comment by basil00** [comment]
The error is `FWP_E_ALREADY_EXISTS` which means `An object with that GUID or LUID already exists.`

I am not sure the cause so you will need to supply more information.

**Comment by gocict** [comment]
Excuse me, what information i can supply?

**Comment by basil00** [comment]
Preferably a step-by-step guide on how to reproduce the error, including what program/code you used.  Whether the error occurs with the example programs.  Whether the error is repeatable or once off.  Whether the error goes away after reboot.  Anything else you think of.

All you posted is an error code and an OS.  If I cannot reproduce the problem it'd be unlikely that it will be fixed.

**Comment by gocict** [comment]
Sorry for the late reply!
I use CitadelCore of TechnikEmpire. It works very well, but on a machine running Windows 10 LTSC x64, it has that error.

**Comment by basil00** [comment]
Unfortunately there is still not much information to go on.  Perhaps @TechnikEmpire has some insight.

Does the problem persist for the WinDivert sample programs, e.g., `netdump.exe`, `streamdump.exe`, etc. included in the [WinDivert 2.0.0-rc release](https://github.com/basil00/Divert/releases)?

**Comment by TechnikEmpire** [comment]
I think the windivert version probably most relevant. All citadecore.windows does is open a pretty straight forward diversion handle.

I wonder if the GUID is one of the unique identifiers for windivert layers/sublayers/filters.

Anyway I don't know what more I could add here. Citadelcore.windows is moved to closed development now and I'm in the process of plugging in a custom written wfp callout driver where I stuff all of my app logic into kernel space.

**Comment by basil00** [comment]
> I wonder if the GUID is one of the unique identifiers for windivert layers/sublayers/filters.

Yes, it must be, although I've never seen this error before nor can see how it could occur.  What version of WinDivert does the public version of CitadelCore use?

> plugging in a custom written wfp callout driver where I stuff all of my app logic into kernel space.

It will be interesting to see what the performance gain is over WinDivert.

**Comment by TechnikEmpire** [comment]
Windivert I think answers a much broader and dynamic range of requirements than I need is all. I only swap packets to be inbound, so fortunately I don't need be as masterful with wfp as you to get my simple job done. Haha

Windivert has been great to develop against and again thanks for making and publishing it.

**Comment by TechnikEmpire** [comment]
Public version uses 1.4.x not 2.x.

**Comment by basil00** [comment]
> I only swap packets to be inbound

For this the `ALE_CONNECT_REDIRECT` layer might be useful, although I've not tried it yet.

**Comment by TechnikEmpire** [comment]
Yeah but I still need to do dynamic inspection of all packets to catch things like socks proxies and tunnels so I'm basically just copying and pasting citadelcore's diverter logic. AFAIK the ale redirect just targets specific static conditions but I could be wrong.

**Comment by TechnikEmpire** [comment]
Anyway on topic of the ticket I would make sure the OP hasn't manually upgraded to windivert 2.x and doesn't somehow have simultaneous versions of windivert or conflicting versions, then lastly I'd make sure that version of windows has all the proper subsystems to even run such a driver.

**Comment by gocict** [comment]
Yes, i'm using Windivert 1.4

**Comment by TechnikEmpire** [comment]
This is a mystery indeed. I don't think its anything to do with citadel. Citadel just asks windivert to give it outbound packets and then redirects them to itself, as it's a thin proxy around a Microsoft web server.

I would do was basil said and run windivert 2.0's passthru example.

I would go even further and run 1.4's passthru as well. Report both results back here.

**Comment by basil00** [comment]
Funnily enough I also got this error after calling `TerminateProcess` on Tallow.  The error occurred when Tallow was relaunched, but the error when away after Tallow was launched a second time.  I have not been able to repeat since, so am not sure of the underlying cause.

**Comment by molenzwiebel** [comment]
Hey, we're seeing this happen for a nonzero amount of users in [snoww/loa-logs](https://github.com/snoww/loa-logs). We noticed that it primarily happens for people that have the NordVPN application running, even if they aren't actually connected with a VPN. Nord ships with a WFP filter called `NordDivert102.2_<name>` which I suspect _might_ be interfering?

Is there any way in which we would be able to get compatibility with that from our end? Perhaps us downgrading to the version that Nord seems to have based theirs on would work?

**Comment by TechnikEmpire** [comment]
According to the examples, this error should be ignored because of conditions where you possibly have a broken state:

https://learn.microsoft.com/en-us/windows/win32/fwp/installing-a-provider

Did your program previously crash or get terminated (perhaps while debugging)? If other driver authors are copying the same GUID that's in WinDivert, well, that's the kind of stupidity that there's no cure for, so I don't think that's the case.

Maybe WinDivert should be updated to treat this error the same way Microsoft does. The logic is sound.

When this error arises, try manually sending an SC terminate to the WinDivert driver service. I assume there's teardown logic in that handler to unregister everything. If after doing that termination, the error is resolved, then we know this is the issue and WinDivert registration code should ignore this error.

**Comment by molenzwiebel** [comment]
It's not that the WinDivert service is somehow in a weird state; it's that I cannot start a WinDivert session at all if Nord's NordDivert is running. Trying to do so results in the `FWP_E_ALREADY_EXISTS`. This happens even if WinDivert isn't loaded and no WinDivert service is running prior to starting my application. As soon as I kill the NordVPN service process, the error goes away.
